### PR TITLE
add simple sanity checks for server startup

### PR DESCRIPTION
**Pull request by serxa**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
added sanity checks on server startup (available memory and disk space, max thread count, etc)

### CODE REVIEW

**Comment by serxa** [comment]
this review make several checks from the list mentioned in #18794

File: programs/server/Server.cpp [hunk 1]
```diff
@@ -516,6 +517,58 @@ void checkForUsersNotInMainConfig(
...
 
+static String readString(const String& path)
+{
+    ReadBufferFromFile in(path);
+    String contents;
+    readStringUntilEOF(contents, in);
+    return contents;
+}
+
+static int readNumber(const String& path)
```
Comment (alesapin) [inline programs/server/Server.cpp:528]:
```suggestion
static int readNumber(const String & path)
```

File: programs/server/Server.cpp [hunk 2]
```diff
@@ -516,6 +517,58 @@ void checkForUsersNotInMainConfig(
 #endif
 }
 
+static String readString(const String& path)
```
Comment (alesapin) [inline programs/server/Server.cpp:520]:
```suggestion
static String readString(const String & path)
```

File: programs/server/Server.cpp [hunk 3]
```diff
@@ -516,6 +517,58 @@ void checkForUsersNotInMainConfig(
...
+
+    if (readNumber("/proc/sys/kernel/pid_max") < 30000)
+        server->context()->addWarningMessage("Linux max PID is too low.");
+
+    if (readNumber("/proc/sys/kernel/threads-max") < 30000)
+        server->context()->addWarningMessage("Linux threads max count is too low.");
+
+    std::string devId = getBlockDeviceId(data_path);
+    if (getBlockDeviceType(devId) == BlockDeviceType::ROT && getBlockDeviceReadAheadBytes(devId) == 0)
+        server->context()->addWarningMessage("Rotational disk with disabled readahead is in use. Performance could be degraded.");
```
Comment (alesapin) [inline programs/server/Server.cpp:559]: `Performance can be degraded`?

File: src/Common/filesystemHelpers.h [hunk 4]
```diff
@@ -18,6 +18,27 @@ using TemporaryFile = Poco::TemporaryFile;
 bool enoughSpaceInDirectory(const std::string & path, size_t data_size);
 std::unique_ptr<TemporaryFile> createTemporaryFile(const std::string & path);
 
+#if !defined(__linux__)
+[[noreturn]]
+#endif
+String getBlockDeviceId(const String & path);
```
Comment (alesapin) [inline src/Common/filesystemHelpers.h:24]: Looks cool. Maybe add some comments?

File: src/Common/filesystemHelpers.h [hunk 5]
```diff
@@ -18,6 +18,27 @@ using TemporaryFile = Poco::TemporaryFile;
...
+
+enum class BlockDeviceType {
+    UNKNOWN = 0,
+    NONROT = 1,
```
Comment (alesapin) [inline src/Common/filesystemHelpers.h:28]: What does it mean?

**Review by alesapin — APPROVED** [review]

File: programs/server/Server.cpp [hunk 6]
```diff
@@ -516,6 +517,58 @@ void checkForUsersNotInMainConfig(
...
+static void sanityChecks(Server* server)
+{
+    std::string data_path = getCanonicalPath(server->config().getString("path", DBMS_DEFAULT_PATH));
+    std::string logs_path = server->config().getString("logger.log", "");
+
+#if defined(OS_LINUX)
+    if (readString("/sys/devices/system/clocksource/clocksource0/current_clocksource").find("tsc") == std::string::npos)
+        server->context()->addWarningMessage("Linux is not using fast TSC clock source. Performance could be degraded.");
+
+    if (readNumber("/proc/sys/vm/overcommit_memory") == 2)
```
[the remark this answers is not in the document]
Comment (serxa) [reply programs/server/Server.cpp]: It looks like we should avoid any unhandled exceptions here to be able to start server anyways

**Comment by alesapin** [comment]
Waiting only for builds, style check unrelated.

**Comment by alesapin** [comment]
@mergifyio update

**Comment by alesapin** [comment]
Failures look unrelated but let's check one more time.

**Pull request merged**